### ISSUE TITLE
`no-test-callback` complains in non Promise based test

### ISSUE THREAD

**Issue by screendriver** [jest-community/eslint-plugin-jest#494]
Following code snippet will ESLint complain with a `error  Illegal usage of test callback  jest/no-test-callback` despite I'm not using a Promise based test:

```ts
it('should something', done => {
  const result = doSomething();
  setTimeout(() => {
    expect(result).toBe(false);
    done();
  }, 0);
});
```

**Comment by SimenB** [comment]
Yes, this rule is meant to discourage _all_ usage of the `done` callback. You can write this instead:

```js
it('should something', () => {
  const result = doSomething();

  return new Promise(resolve => {
    setTimeout(() => {
      expect(result).toBe(false);
      resolve();
    }, 0);
  });
});
```

Note that this test will time out if `expect` fails as the promise (or callback in your original example) is never rejected.

A better version would be something like this

```js
it('should something', async () => {
  const result = doSomething();

  await new Promise(resolve => setTimeout(resolve, 0));

  expect(result).toBe(false);
});
```

I realise your code is slimmed down example code, but if you have a callback based api using https://nodejs.org/api/util.html#util_util_promisify_original or a manual `await new Promise((resolve, reject) => callBackApi((err, res)=> err ? reject(err) : resolve(res)})` is a safer way to do async tests

**Comment by screendriver** [comment]
> this rule is meant to discourage all usage of the done callback

Ok, this is new for me. Well the **all** part of it. So I will go with this one

```ts
await new Promise(resolve => setTimeout(resolve, 0));
```

:+1:

**Comment by nickjohngray** [comment]
for me the code below is clearer  

it('should do it', () => new Promise((done) => {
      mount(props);
      action();
      setTimeout(() => {
        action2();
        done();
        expect(xxx).toBeCalled();
      }, 300);
    }));
